### ISSUE TITLE
"unique" attribute ignored in model validation (again)

### ISSUE THREAD

**Issue by kenjinsama** [balderdashy/sails-mongo#193]
Hi,

After searching a looong time and look over some solutions whithout succes, i resigned to post an issue for this problem.

**Let me explain:** I'm programming an authentification system over sails and for wich I use sails-mongo adapter. For the past weeks all seemed to be working. But today I faced a big problem: the **"unique"** attribute was not working and the error was triggered by mongoDB and not by the waterline validation.

**EDIT 1:**
It seems that sails never check the unicity in mongoDB (no query registered by the logs before "the insertion", only mines).

### The Error:

> Error (E_UNKNOWN) :: Encountered an unexpected error
> MongoError: insertDocument :: caused by :: 11000 E11000 duplicate key error index: yasn.participant.$pseudo_1  dup key: { : "kenjin" }
>     at Object.toError (/home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/utils.js:114:11)
>    at /home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/collection/core.js:111:29
>    at /home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/db.js:1132:7
>    at /home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/db.js:1846:9
>    at Server.Base._callHandler (/home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/base.js:445:41)
>    at /home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/server.js:468:18
>    at MongoReply.parseBody (/home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
>    at null.<anonymous> (/home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/server.js:426:20)
>    at emit (events.js:95:17)
>    at null.<anonymous> (/home/stephane/Documents/yasn/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:201:13)
>
> Details:  MongoError: insertDocument :: caused by :: 11000 E11000 duplicate key error index: > yasn.participant.$pseudo_1  dup key: { : "kenjin" }

Before any comment, i would like to say that i've read tons of issues for that problem and i've already checked my sails environnement (like my waterline is as expected to test on **"drop"** mode). I know there is solution like do the nasty job by myself bofore the validation (I will do but i'm very lazy so i hope the problem has or will have a solution as soon as it is possible :dart: ).
If I'm missing something i can reproduce the issue as much as I want (sadly) and give you other details. I will give you some of my configuration files at first. If i'm missing something basic, shame on me :+1: 

###### Profil.js

```javascript
/**
* Profil.js
*
* @description :: TODO: You might write a short summary of how this model works and what it represents here.
* @docs        :: http://sailsjs.org/#!documentation/models
*/

module.exports = {

    types: {

        /* Vérifie que l'avatar est bien Aang, bon, en fait ça vérifie si on */
        /* a bien un url ou un uri de type image                             */

        avatar: function (avatar) {
            var is_url = avatar.match(/^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/|ftp:\/\/)[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/i);
            var is_uri = avatar.match(/data:image\/.{1,4};base64,/);

            console.log(is_url + ' - ' + is_uri + ' - ' + (is_url || is_uri))
            return (is_url || is_uri);
        },

        /* Le pseudo doit pouvoir appartenir au participant qui est */
        /* actuellement connecté en plus d'être unique */

        isMine: function (pseudo) {
            console.log((!this.pseudo_found || pseudo === this.session.pseudo) + ' - ' + pseudo + ' + ' + this.session.pseudo);
            return (!this.pseudo_found || pseudo === this.session.pseudo);
        },

        /* Vérifie qu'on peut encore créer un profil avec ce compte */

        limited: function (email) {
            return (this.is_over_limite);
        }
    },

    attributes: {
        pseudo: {
            type: 'alphanumeric',
            required: true,
            minLength: 3,
            maxLength: 16,
            isMine: true,
            unique: true
        },
        civilite: {
            type: 'string',
            in: ['H', 'F'],
            required: true
        },
        prenom: {
            type: 'string',
            required: true,
            maxLength: 32
        },
        nom: {
            type: 'string',
            required: true,
            maxLength: 32
        },
        date_de_naissance: {
            type: 'datetime',
            required: true
        },
        adresse: {
            type: 'string',
            required: true,
            maxLength: 128
        },
        code_postal: {
            type: 'string',
            required: true,
            maxLength: 12
        },
        ville: {
            type: 'string',
            required: true,
            maxLength: 128
        },

        /* Récupéré depuis la variable de session */

        email: {
            type: 'email',
            required: true
        },

        /* Taille maximum autorisée 192 ko mais en base 64: 263 ko */

        avatar: {
            type: 'string',
            maxLength: 269353
        }
    },

    /* On effectue les transformations nécessaires à la validation (date de naissance) */

    beforeValidation: function(doc, cb) {
        doc.date_de_naissance = new Date(doc.date_de_naissance);

        /* On vérifie l'unicité du pseudo dans la deuxième collection */

        Participant.count({pseudo: doc.pseudo}).exec(function (err, found) {
            if (err) console.log(err);
            doc.pseudo_found = found;
            
            /* On vérifie le nombre de profils du participant */

            Profil.count({email: doc.email}).exec(function (err, found) {
                console.log(found);
                if (err) console.log(err);
                doc.is_over_limite = found >= doc.session.nb_profil_max
                cb();
            });
        });
    },

    /* On effectue les transformations nécessaires au stockage (geolocalisation) */

    beforeCreate: function(doc, cb) {

        /* On efface ce qui ne sera pas stocké en bdd */

        delete doc.session;
        delete doc.is_over_limite;
        delete doc.pseudo_found;   

        if (!doc.geo_loc)
        {
            sails.maps.geocode(doc.adresse + ',' + doc.code_postal + ',' + doc.ville, function (err, data) {
                if (err) cb(err);
                if (data.results.length > 0)
                {
                    doc.geo_loc                 = {};
                    doc.geo_loc.type            = 'Point';
                    doc.geo_loc.coordinates     = [];
                    doc.geo_loc.coordinates[0]  = data.results[0].geometry.location.lng;
                    doc.geo_loc.coordinates[1]  = data.results[0].geometry.location.lat

                    cb();
                }
                else
                {
                    cb();
                }
            });
        }
        else
        {
            cb();
        }   
    }
};
```

###### Participant.js

```javascript

module.exports  = {

    types: {

        /* Si tu n'as pas compris à quoi sert cette fonction tu es viré */

        confirmation: function (confirmation) {
            return (this.mot_de_passe === confirmation);
        },

        /* Le pseudo doit être unique dans les deux collections */

        notInProfil: function (pseudo) {
            return (!this.pseudo_found);
        }
    },
    attributes: {
        pseudo: {
            type: 'alphanumeric',
            required: true,
            minLength: 3,
            maxLength: 16,
            notInProfil: true,
            unique: true
        },
        mot_de_passe: {
            type: 'string',
            minLength: 6,
            maxLength: 64,
            required: true
        },
        confirmation: {
            type: 'string',
            minLength: 6,
            maxLength: 64,
            required: true,
            confirmation: true
        },
        email: {
            type: 'email',
            required: true,
            unique: true
        },
        nb_profil_max: {
            type: 'integer',
            defaultsTo: 2
        }
    },

    beforeValidation: function(doc, cb) {

        /* On vérifie l'unicité du pseudo dans la deuxième collection */

        Profil.count({pseudo: doc.pseudo}).exec(function (err, found){
                if (err) console.log(err);
                doc.pseudo_found = found;
                cb();
        });        
    },

    beforeCreate: function(doc, cb) {
        doc.mot_de_passe    = sails.sum(doc.mot_de_passe);
        doc.id_validation   = sails.hashify(doc);
        
        /* On efface ce qui ne sera pas stocké en bdd */

        delete doc.confirmation;
        delete doc.captcha;
        delete doc.pseudo_found;

        cb();    
    }
};
```

**Comment by kokujin** [comment]
+1

**Comment by leobudima** [comment]
+1

**Comment by mike360** [comment]
+1 for this. I think even if the error still comes from the database as opposed to a validation function, it should fail gracefully with a 400 error rather than a 500 server error and dumping the error to console.

**Comment by douglassantos** [comment]
+1

**Comment by IanVS** [comment]
+1

**Comment by Oldwo1f** [comment]
+1 I got the same problem.

Do i have to make a manual verification on my uniques field before inserting db?

**Comment by ArtworkAD** [comment]
@Oldwo1f yes. Since the unique option is not working you can do this manually inside the boostrap.js file.
```
Foobar.native(function(err, collection) {
  collection.ensureIndex('name', {
    unique: true
  }, function(err, result) {
    if (err) {
      sails.log.error(err);
    }
  });
});
```

If you try to create a model which name already exists, mongodb will send a dublicate key error and sails responds with http status 500.

----

@mike360  I agree that sails should process database errors and return proper error codes and not just 500.

**Comment by Oldwo1f** [comment]
Thanks for you answer.
I used another way to ensure unique attribute:

- Before inserting a user in db i query user by email and by user name with async.parraralle
```javascript
async.parallel([
		    function(callback){
		    		User.findOne({email: user.email }).exec(function (err,result){
			            if(err){
			            	callback(null, 'emailOk');						
						}else{
							if(typeof(result) !='undefined')
							{	
								callback('emailError');
							}else{
								callback(null, 'emailOk');
							}
						}
					});
		    },
		    function(callback){
		            User.findOne({pseudo: user.pseudo }).exec(function (err,result){
		            	if(err){
			            	callback(null, 'pseudoOk');
						}else{
							if(typeof(result) !='undefined')
							{	
								callback('pseudoError');
							}else{

								callback(null, 'pseudoOk');
							}
						}
					});
		    }
		],
		function(err, results){
			if(err){
				res.status(401).send(err);
			}
			else{
				console.log('CREATE USER');
	}
});
```

Can you explain a bit more your method within the bootstrap.js file. I think my code is a bit dirty and i would like to understand your solution.

Thanks

**Comment by ArtworkAD** [comment]
@Oldwo1f sure. In your special case add following code to **config/bootstrap.js**
```
User.native(function(err, collection) {
  collection.ensureIndex({'email', 'pseudo'}, {
    unique: true
  }, function(err, result) {
    if (err) {
      sails.log.error(err);
    }
  });
});
```

This grabs the underlaying mongodb node driver (bypassing waterline) and ensures indices for email and pseudo property of your User collection and marks them as unique.

So when you try to set a value that already exists, either by creating or updating a user, mongodb will raise an error and the error will be returned by sails with a http 500 error code.

**Comment by ArtworkAD** [comment]
Just found out that a nice solution would be to create a policy. E.g. **beforeChangeUser.js** and add the policy like this:

    UserController: {
        create: ['beforeChangeUser'],
        update: ['beforeChangeUser']
    }

Then make your checks inside the policy.

**Comment by mlagace** [comment]
I just re-build my unique indexes in my database and it works fine after that.

**Comment by Kikobeats** [comment]
@mlagace how?

This issue is frustrating :(

**Comment by mlagace** [comment]
@Kikobeats In mongo, using the mongo shell, you can just add your unique indexes manually for the model you want. e.g. 

    > db.user.ensureIndex({'email': 1}, {unique: true}) for a simple unique index.

 If you have multiple unique constraints, then you need to **ensureIndex** for each unique key. Once you add all your indexes lift your app and it should work as intended now. You can also add compound indexes with the following command:

    > db.user.ensureIndex({'email': 1, 'company': 1})

**Comment by Kikobeats** [comment]
@mlagace it's true, works! why not sails-mongo do this by default?

**Comment by mlagace** [comment]
That's a good question @Kikobeats , I think unique indexes should be added automatically even if you use *migrate: safe*. Not sure why it doesn't do it though. For now you have 2 solutions:
- Add your unique indexes manually
- Use @ArtworkAD's solution above which will also work great. I think this would be the best solution IMO. That way, every time you deploy you would automatically ensure your indexes. **WARNING:** Using this solution you can put a lot of stress on your memory if you have non unique indexes to add as well  since indexes use up a lot of RAM to be effective.

**Comment by dpoellath** [comment]
+1

workaround: add index for attribute (with unique flag)

**Comment by victorbadila** [comment]
+1 I think this feature is very important from a mature framework's point of view

On the other hand @mlagace I have rebuilt the indexes in my db just like in your example, however what were you referring to when you said "your app and it should work as intended now"? The error I get when trying to insert/update duplicate values is still the one generated by mongo.

@ArtworkAD Regarding your policy solution, it works, however if you have multiple unique fields and want to know what fields in the new objects are duplicated (for example in order to notify a client with exact details) I think you would have to make separate checks against the db and I don't know how this would be performance wise.

**Comment by dogrocker** [comment]
I think this is very important too. If the unique can't work Docs need to tell that the developer can be find the solution to handle them.

**Comment by gilesbutler** [comment]
+1 for this as well @ArtworkAD's solution works but the user doesn't get any sort of helpful error message unfortunately

**Comment by galcivar** [comment]
+1

**Comment by devinivy** [comment]
Could someone write a failing test for this?  That would be a huge step forward!

**Comment by praveenaj** [comment]
I solved it by following this method:
http://stackoverflow.com/questions/22868420/how-do-i-handle-a-unique-field-in-sails

**Comment by mmiller42** [comment]
Hello guys. I know this is far from the best answer and a really hackish solution, but this solution was quickest for my situation. This applies only to Sails, and would have to be implemented differently if you're using Waterline independently.

Basically, the blueprint actions (which can be seen in `node_modules/sails/lib/hooks/blueprints/actions/`) `create` and `update` both call `res.negotiate` when they get an error from whatever Waterline adapter you're using. I simply modified `res.negotiate` to parse the error and adjust the code from 500 to 400 if it turned out to be a MongoDB duplicate key error. Again, really hackish, but if you have multiple models that have unique constraints and/or you don't want to create policies that run extra queries that could be vulnerable to race conditions, this seems to do the trick, and it doesn't require messing with sails-mongo, sails, or waterline (none of which I'm comfortable enough yet as a developer to contribute to).

1. Copy `node_modules/sails/lib/hooks/responses/defaults/negotiate.js` into your `/api/responses/` directory to override the default `res.negotiate` method.

2. After the try/catch block but before the if statusCode blocks, paste this in:

```js
  // MongoDB duplicate key error code
  if (err.originalError.name === 'MongoError' && err.originalError.code === 11000) {
    statusCode = 400;

    // Example errmsg: `E11000 duplicate key error index: db_name.model_name.$attribute_name_1 dup key: { : "value" }`
    var model = /: (.*?)\.(.*?)\./.exec(err.originalError.errmsg)[2];
    model = model.charAt(0).toUpperCase() + model.substr(1); // capitalize name
    // Create the error to look like Waterline validation errors (probably should use a WLError constructor, but I
    // haven't figured that out yet)
    body = {
      error: 'E_VALIDATION',
      status: 400,
      summary: '1 attribute is invalid',
      model: model,
      invalidAttributes: {}
    };

    var fieldName = /\.\$(.*?)_\d\s/.exec(err.originalError.errmsg)[1]; // Extract field name from message
    body.invalidAttributes[fieldName] = { rule: 'unique', message: '`' + err.originalError.getOperation()[fieldName] + '` is not a unique value' };
  }
```

*Known Limitations:*

* Pretty crude -- if format of MongoDB error messages change, this won't work (but I don't see another way to do this other than with regular expressions)
* The `model` property will only be the name of your model if the name of your model = the name of the Mongo collection, capitalized. If you override the default table name, this won't work.
* The same applies to if the name of your attribute isn't the same as Mongo's (does this happen? I don't know) or has anything really weird in it.
* If Waterline allows you to configure your index name, this won't work. It depends on an index name being `{name_of_your_attribute}_{a number}`.
* Subject to breaking if the format of a Waterline validation error changes. This should be changed to use a `WLValidationError` instance instead of emulating it, but I'm not sure how.
* If you have a plugin to beautify Waterline validation errors, I'm afraid this will probably not work, since this error is generated right before sending it to the client. I'm not sure though, haven't tested that.

Anyone with ideas for improvement, please let me know, I'd love to make this a little more robust and not so hacky.

**Comment by mmiller42** [comment]
I was concerned with the other recommended solutions in this thread, since I think the insertion operation should be atomic. (1) The policy implementation is vulnerable to race conditions: if two requests are sent simultaneously, there is no guarantee that the 500 error would be thrown for one client; and (2) It seems inefficient to make two trips to the database for an insertion.

My implementation isn't perfect or standard enough to create a pull request in `sails-mongo`, since it is a little hackish in how it works (it requires using regular expressions to parse the error and extract the relevant data: collection name, index name, and value), as well as that it makes some unsafe assumptions (model name = capitalized collection name, attribute name = index name without the tailing _digits).

I've created a fork of `sails-mongo` that produces a `WLValidationError` instead of a raw Mongo error when it encounters a duplicate key error. You can install it by modifying your `package.json` like so:

```js
"sails-mongo": "https://github.com/mmiller42/sails-mongo/archive/22ed63ad6760f69117e5fd019c6d7a05a69864be.tar.gz",
```

followed by a `npm install`.

Again, please note that there is no guarantee that the value of the `model` property reflects the actual name of your model, if your model name is different from your collection name; and similarly, if the name of the index is not the name of an attribute followed by an _ and some digits, the attribute name in the error won't be extracted correctly.

**Comment by gayanhewa** [comment]
hey guys , any fix for this ? I still find it problematic when using blueprint generated api methods with models that has unique keys. Difficult to figure out a proper solution instead of hacking the models to work around. Resulting in having to patch the models each time, trying to look for better work around to get things done. Any ideas appreciated

**Comment by particlebanana** [comment]
Ok so there are two issues here. The main issue it seems is that a validation error isn't being returned and instead the mongo error is coming back. If someone could create an issue for that to track it that would be helpful.

This requires an inspection of the error inside the mongo adapter and then it can be formatted to a proper validation error. Something similar is done on the [postgresql adapter](https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L1225-L1247). This is reasonable and is something we could add easily.

The second issue is that you want Waterline to edit your database but only for certain cases. In the SQL world this is where migrations come into play. If you need to add an index or a column you create a migration and run it. You can do this in development/production etc.

In mongo migrations are not needed. So as you write data it saves the document on demand. If you need to add an index you use the `ensureIndex` command as posted above. The auto-migrations feature of Sails/Waterline is really dumb in that when it runs in `drop` or `alter` mode it will automatically add these indexes but in `safe` mode which is turned on by default in production it won't touch anything in your database on start.

So i'm open to solutions on how to solve this. I want to make sure we have a mode that won't alter a production database but I guess we also need a way of setting up your production database in the no-sql adapters.

**Comment by mmiller42** [comment]
@gayanhewa @particlebanana I submitted a PR to convert unique constraint errors to validation errors here, if anyone cares to review and critique: https://github.com/balderdashy/sails-mongo/pull/319

**Comment by particlebanana** [comment]
@mmiller42 yeah that looks great!

**Comment by gayanhewa** [comment]
@mmiller42 your PR seems to be in par with how its handled with the other adapters. :+1:

**Comment by particlebanana** [comment]
Ok thanks to @mmiller42 the first issue is now solved in `0.11.3`. The second issue still exists though.

**Comment by sailsbot** [comment]
Thanks for posting, @kenjinsama.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by OwaysisApp** [comment]
this is still an issue^

**Comment by mmiller42** [comment]
Could you paste in the error you receive from MongoDB? It's possible it isn't matching the regular expression, there seems to be some variance in the error returned from Mongo.
